File: 백준/Silver/1417. 국회의원 선거/국회의원 선거.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;

// https://www.acmicpc.net/problem/1417
public class Main {
    private static int N; // 후보의 수
    private static int D; // 다솜이
    private static PriorityQueue<Integer> pq;


    public static void main(String[] args) throws IOException {
        init();

        int minCnt = getMinCnt();
        System.out.print(minCnt);
    }//main

    
    private static int getMinCnt() {
        int cnt = 0;

        while(!pq.isEmpty()) {
            if(pq.peek() < D) break;

            pq.offer(pq.poll() - 1);
            D++;
            cnt++;
        }

        return cnt;
    }//getMinCnt

    
    private static void init() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine()); // 후보의 수
        D = Integer.parseInt(br.readLine()); // 다솜이

        pq = new PriorityQueue<>(Comparator.reverseOrder());

        for(int i=1; i<N; i++) {
            int cnt = Integer.parseInt(br.readLine());
            pq.offer(cnt);
        }

        br.close();
    }//init

    
}//class

File: 백준/Silver/1417. 국회의원 선거/국회의원 선거.kt
import java.util.PriorityQueue

/**
 * https://www.acmicpc.net/problem/1417
 * 다솜이가 매수해야하는 사람의 최솟값을 출력하는 프로그램을 작성
  */

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt() // 후보 수
    var d = readLine().toInt() // 다솜이의 표(다솜이는 기호 1번)
    val pq = PriorityQueue<Int>(compareByDescending { it }) // 최대힙
    var minCnt = 0 // 다솜이가 매수해야하는 사람의 최솟값

    repeat(n-1) {
        val cnt = readLine().toInt()
        pq.offer(cnt)
    }

    while (pq.isNotEmpty() && pq.peek() >= d) {
        val top = pq.poll() - 1
        
        pq.offer(top)
        d++
        minCnt++
    }

    close()
    print(minCnt)
}
